### PR TITLE
Add handling of timeouts to `ApiWebRequest`

### PR DESCRIPTION
**Pull request by andrewlock**
## Summary of changes

Fixes timeout handling for `ApiWebRequest`

## Reason for change

Basically, timeouts never worked in our `ApiWebRequest`, because the APIs we call (`GetRequestStreamAsync`/`GetResponseAsync`) simply don't pay attention to it. That means we effectively had no timeout on any http calls for .NET Framework.

## Implementation details

_Actually_ handle a timeout in `ApiWebRequest`. Technically this still isn't _quite_ the same as our `HttpClient` implementations, because for `HttpClient` we're using a mode that buffers everything in memory immediately, and applies the timeout including that. That's more effort to do with `ApiWebRequest`, as you need to explicitly pass the cancellation token through to the _response_ etc, which all becomes kinda different to `HttpClient`. Ultimately, what this adds is better than what we have _today_

**Note**: My assumption is that we will _not_ want to buffer everything in memory soon, seeing as we're always explicitly copying the body or stream decoding it. Today, we are doing more copying than we should. I expect we'll look at using `HttpCompletionOption.ResponseHeadersRead` instead soon, in which case, we will need to unify our handling of timeouts in any case. A follow up process.

## Test coverage

Added unit tests for the timeouts (which hopefully are _not_ flaky), given I'm cheating by passing in a deterministic cancellation token instead of trying to "wait for it to fail". There's a bunch of edge cases around exactly _when_ the timeout is cancelled which we handle. There's the occasional additional edge case that we can't test without a lot more work, so left that to manual testing instead

## Other details

Stacked on:
- https://github.com/DataDog/dd-trace-dotnet/pull/9145

Discovered as part of:
- https://github.com/DataDog/dd-trace-dotnet/pull/9042

### CODE REVIEW

**Comment by andrewlock** [comment]
@codex review

File: tracer/src/Datadog.Trace/Agent/Transports/ApiWebRequest.cs [hunk 1]
```diff
@@ -184,12 +184,32 @@ private void ResetRequest(string method, string? contentType, string? contentEnc
...
+                // exceptions -- nothing observes them there.
+
+                // Note that this deadline is currently _only_ scoped to the "send": connect,
+                // writing the request body, and waiting for response headers - ending when
+                // SendAsync returns. Reading the response body afterwards is not covered.
+                //
+                // This diverges slightly from HttpClient, where the timeout covers the _whole_ read
+                // currently, however, this _may_ change if we switch to using HttpCompletionMode.
+                // ResponseHeadersRead to avoid the full buffering into memory as we do today.
+                cts.Token.Register(static s => { try { ((HttpWebRequest)s!).Abort(); } catch { } }, _request);
```
[the remark this answers is not in the document]
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Agent/Transports/ApiWebRequest.cs]: fixed

[the remark this answers is not in the document]
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Agent/Transports/ApiWebRequest.cs]: fixed

**Comment by andrewlock** [comment]
@codex review

File: tracer/test/Datadog.Trace.Tests/Agent/Transports/ApiRequestTests.cs [hunk 2]
```diff
@@ -43,14 +47,193 @@ public ApiRequestTests(ITestOutputHelper output)
...
+                    const int chunkSize = 64 * 1024;
+                    for (var offset = 0; offset < bytes.Length; offset += chunkSize)
+                    {
+                        var count = Math.Min(chunkSize, bytes.Length - offset);
+                        await stream.WriteAsync(bytes, offset, count);
+                    }
+                },
+                cts);
+
+            await listener.Accepted;
```
Comment (NachoEchevarria) [inline tracer/test/Datadog.Trace.Tests/Agent/Transports/ApiRequestTests.cs:225]: listener.Accepted means the server accepted the TCP connection but we dont know if the request-body producer has started, finished, or is blocked writing. It would need to establish that a write has started and remains incomplete before cancellation.

**Review by NachoEchevarria — COMMENTED** [review]
Left one minor comment. Thanks!
